tools/mason/MasonInit.chpl

```diff
@@ -0,0 +1,160 @@
+      if arg == '' {
+        name = arg;
+      } 
+      else {
+        name = arg;
+      }
+    }
+    //checks if path is given as parameter or not
+    if name == '' {
+      if isFile("Mason.toml") {
+        var projectName = ""; 
+        var version = "";
+        const toParse = open("Mason.toml", iomode.r);
+        const tomlFile = new owned(parseToml(toParse));
+        if (tomlFile["brick"]["name"]) {
+          projectName = tomlFile["brick"]["name"].s;
+        }
+        else {
+          throw new owned MasonError("Project Name doesn't exist in Mason.toml");
+        }
+        if (tomlFile["brick"]["version"]){
+          version = tomlFile["brick"]["version"].s;
+        } 
+        else {
+          throw new owned MasonError("Version doesn't exist in Mason.toml");
+        }
+        validateAndInit(projectName, vcs=true, show=false);
+        checkVersion(version);
+      } 
+      //If Mason.toml not present, proceed with init
+      const cwd  = getEnv("PWD");
+      const currDir = basename(cwd);
+      var name = currDir;
+      var path = '.';
+      makeBasicToml(name,path);
+      ValidateInit(path);
+      writeln("Initialized new library project: " + name);
+    } 
+    else {
+      // if the target directory in path doesnt exist, throw error
+      // if target directory exists, check for files && validate
+      // create folders and toml file without overwriting anything 
+      // if TOML file exists, check for values in it and validate
+      var path = name;
+      if(isDir(path)) {
+        ValidateInit(path);
+        writeln("Initialized new library project in " + path + ": " + basename(path));
+      } 
+      else {
+        throw new owned MasonError("Directory does not exist:" + path +
+                              ". Did you mean 'mason new' to create a new project from scratch?"); 
+      }
+    }
+  }
+  catch e: MasonError {
+    writeln(e.message());
+    exit(1);
+  }
+}
+
+proc ValidateInit(path: string) throws {
+  var files = [ "/Mason.toml" , "/src" , "/test" , "/example", "/.git", ".gitignore" ];
+  var ToBeCreated : list(string);
+  for idx in 1..files.size do {
+    var file = files(idx);
+    if file == "/Mason.toml" {
+      if isFile(path + file) {
+        //scan using TOML reader and validate
+        const toParse = open(path + file, iomode.r);
+        const tomlFile = new owned(parseToml(toParse));
+        var projectName = tomlFile["brick"]["name"].s;
+        var version = tomlFile["brick"]["version"].s;
+        validateAndInit(projectName, vcs=true, show=false);
+        checkVersion(version);
+      } 
+      else {
+        ToBeCreated.append(file);
+      }
+    } 
+    else {
+      var dir = file;
+      if isDir(path + dir) == false {
+        ToBeCreated.append(dir);
+      }
+    }
+  }
+  if ToBeCreated.size == 1 {
+    throw new owned MasonError("Library project has already been initialised.");
+    exit(1);
+  }
+
+  for file in ToBeCreated{
```

Comment:
I am not a huge fan of the `makeProjectFiles()` refactor, due to the reliance on special string values. What do you think about the following strategy?

Create a set of functions in `MasonNew` that create each component of the package:

```chpl
// Already exists
proc makeBasicToml(name, path=name)
proc gitInit(name) 
proc addGitIgnore(name)

// to add
proc makeSrcDir(name, path=name)
proc makeModule(name, path=Path.pathJoin(name, 'src'))
proc makeTestDir(name, path=name)
proc makeExampleDir(name, path=name)
```

`makeProjectFiles()` can call each of these individually, while this loop can do something like:

```chpl
const name = basename(path);

for file in ToBeCreated {
  select file {
    when '/Mason.toml' {
       makeBasicToml(name, path);
    }
    when '/.git' {
      gitInit(path, show=false);
    }
   // 
   // continue this pattern for each project file
   //
    otherwise {
      throw new owned MasonError('Unexpected file name to be generated.');
    }
  }
}
```

Reply:
I did think about this too ! But I was trying my best not to modify previously written code in `MasonNew`. Thinking of this, I feel we should consider checking other `Mason` files for refactoring in future issues.